### PR TITLE
Fix error for printing in IJuliaMode

### LINKED ISSUE #2293: Powers print incorrectly in IJulia mode
image

The decimal point is getting mangled.

### PR DESCRIPTION
**Pull request by guilhermebodin**
Fixes #2293

Captura de Tela 2021-01-02 às 02 59 30

### CODE REVIEW

**Comment by joaquimg** [comment]
Does it work for negative exponents?

**Comment by guilhermebodin** [comment]
Probably not

**Comment by guilhermebodin** [comment]
Fixed some more corner cases

Captura de Tela 2021-01-04 às 01 30 15

File: test/print.jl [hunk 1]
```diff
@@ -197,18 +197,39 @@ end
...
         io_test(REPLMode, constr_ge, "sin(x) - 1.0 $ge 0")
         io_test(REPLMode, constr_eq, "sin(x) - 1.0 $eq 0")
         # Note: This is inconsistent with the "x in [-1, 1]" printing for
         # regular constraints.
         io_test(REPLMode, constr_range, "0 $le sin(x) $le 1")
+        io_test(REPLMode, constr_exponent_1, "x ^ 4.0 - 1.0 $le 0")
+        io_test(REPLMode, constr_exponent_2, "x ^ 40.23 - 1.0 $le 0")
+        io_test(REPLMode, constr_exponent_3, "(x ^ 4.0 + x ^ 3.0 + x ^ 2.0) - 1.0 $le 0")
+        io_test(REPLMode, constr_exponent_4, "(x ^ -4.0 + x ^ -3.0 + x ^ -2.0) - 1.0 $le 0")
+        io_test(REPLMode, constr_exponent_5, "x ^ (x * x) - 1.0 ≤ 0")
```
Comment (odow) [inline test/print.jl:218]: Hard-coding the inequality symbol causes CI failures: https://github.com/jump-dev/JuMP.jl/pull/2420/checks?check_run_id=1641836693#step:6:199.

Interpolate the symbol instead as above.

File: src/print.jl [hunk 2]
```diff
@@ -617,10 +617,15 @@ end
...
-    return string(_tape_to_expr(model, 1, c.nd, adjmat(c.nd), c.const_values,
-                                [], [], model.nlp_data.user_operators, false,
-                                false, mode))
+function nl_expr_string(model::Model, print_mode, c::_NonlinearExprData)
+    nl = string(_tape_to_expr(model, 1, c.nd, adjmat(c.nd), c.const_values,
+                              [], [], model.nlp_data.user_operators, false,
+                              false, print_mode))
+    if print_mode == IJuliaMode
+        # Replace exponents x ^ 4.0 for x ^ {4.0}
+        nl = replace(nl, r"(\^) (?<exponent>(?=\()\(.+\)|-?\d+\.\d+)" => s"^ {\g<exponent>}")
```
Comment (odow) [inline src/print.jl:626]: Another option and one maybe more robust, is to work with the Julia expression before converting it to a string.
```Julia
 _latexify(ex) = ex
function _latexify(ex::Expr)
    if ex.head != :call
        return ex
    end
    for i = 1:length(ex.args)
        ex.args[i] = _exponential_to_curly(ex.args[i])
    end
    if ex.args[1] == :^
        ex.args[3] = Expr(:braces, ex.args[3])
    end
    return ex
end

function foo(ex)
    return string(_latexify(ex))
end

foo(:(x ^ (123 + 1)))
```

File: src/print.jl [hunk 3]
```diff
@@ -617,12 +617,37 @@ end
...
+        end
+        if ex.args[1] == :^
+            if isa(ex.args[3], Expr)
+                ex.args[3] = _latexify_exponentials(ex.args[3])
+            end
+            ex.args[3] = Expr(:braces, ex.args[3])
+            return ex
+        end
+    end
+    return ex
```
Comment (odow) [inline src/print.jl:649]: Prefer using dispatch over `isa` checks. You can also drop the `:call` check if you make sure the `.args` has the correct number of elements.
```Julia
_latexify_exponentials(ex) = ex
function _latexify_exponentials(ex::Expr)
    for i = 1:length(ex.args)
        ex.args[i] = _latexify_exponentials(ex.args[i])
    end
    if length(ex.args) == 3 && ex.args[1] == :^
        ex.args[3] = Expr(:braces, ex.args[3])
    end
    return ex
end
```

**Review by odow — CHANGES_REQUESTED** [review]

**Review by odow — APPROVED** [review]
Looks good, assuming tests pass.

**Comment by odow** [comment]
I found a similar problem: https://github.com/jump-dev/JuMP.jl/issues/2422
